haystack/agents/base.py

```diff
@@ -0,0 +1,271 @@
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
+            raise AgentError("Agents without tools cannot be run. Add at least one tool using add_tool().")
+        if params and "max_iterations" in params:
+            remaining_iterations = params["max_iterations"]
+        else:
+            remaining_iterations = self.max_iterations
+
+        transcript = self.prompt_text
+        final_answer = ""
+
+        # In each iteration, choose a tool, generate input for it, and run it until the final answer is found or the
+        # maximum number of iterations is reached.
+        while remaining_iterations > 0:
+            remaining_iterations -= 1
+
+            prompt_template = PromptTemplate("think-step-by-step", transcript, ["query"])
+            pred = self.prompt_node.prompt(prompt_template=prompt_template, query=query)
+            transcript += f"{pred[0]}\n"
+
+            tool_name, tool_input = self._parse_tool_name_and_tool_input(pred=pred[0])
+            if tool_name == "Final Answer":
+                final_answer = tool_input
+                break
+            if tool_name not in self.tools:
+                # TODO we could try to recover from this error by rephrasing the question and retrying
+                raise AgentError(
+                    f'The Agent tried to use a tool "{tool_name}" but the registered tools are only {self.tools.keys()}.'
+                )
+            pipeline_or_node = self.tools[tool_name].pipeline_or_node
+            # We can only pass params to pipelines but not to nodes
+            if isinstance(pipeline_or_node, (Pipeline, BaseStandardPipeline)):
+                result = pipeline_or_node.run(query=tool_input, params=params)
+            else:
+                if isinstance(pipeline_or_node, BaseRetriever):
+                    result = pipeline_or_node.run(query=tool_input, root_node="Query")
+                else:
+                    result = pipeline_or_node.run(query=tool_input)
+
+            observation = ""
+            # if pipeline_or_node is a node it returns a tuple. We use only the output but not the name of the output.
+            if isinstance(result, tuple):
+                result = result[0]
+            if isinstance(result, dict):
+                if "results" in result and len(result["results"]) > 0:
+                    observation = result["results"][0]
+                elif "answers" in result and len(result["answers"]) > 0:
+                    observation = result["answers"][0].answer
+                elif "documents" in result and len(result["documents"]) > 0:
+                    observation = result["documents"][0].content
+                else:
+                    # no answer/document/result was returned
+                    observation = ""
+            transcript += f"Observation: {observation}\nThought: Now that I know that {tool_input} is {observation}, I "
```

Comment:
Don't include 189 in `_extract_observation`; it's important imho to have this line of code here (for readability).